Exercise upgrades with custom robotest images

wadells opened the pull request:
## Description
This PR introduces two notable changes, as well as many smaller ones:

1) Robotest now produces custom images for a variety of gravity versions, and uses these during upgrade testing.
2) Docker permissions issues (#2103) are fixed, allowing Jenkins to properly clean up gravity build directories for the first time in over a year.

## Type of change
* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs
* Closes #1915
* Closes #2103
* Ports #2023
* Depends on https://github.com/gravitational/gravity.e/pull/4336

## TODOs

- [x] Enterprise PR
- [x] Self-review the change
- [x] Write documentation
- [x] Green PR build
- [x] Green Nightly build
- [x] Gather before/after artifact size stats, runtime stats
- [x] Address review feedback

## Testing done
Nightly run: [logs](https://console.cloud.google.com/logs/query;query=severity%3E%3DINFO%0Alabels.__suite__%3D%221501e656-9ca9-45c7-a085-b0a5704a2c99%22;timeRange=2020-09-19T06:09:00.000Z%2F2020-09-19T07:27:00.000Z?authuser=0&project=kubeadm-167321)

This PR's jenkins job shows the tests run to completion on PRs.

Examples of using the Makefile documented in the README.md.

## Additional information

### New Configurations

<details><summary>pr configs</summary>

```
walt@gcp:~/git/gravity$ make -C e/assets/robotest get-config ROBOTEST_CONFIG=pr     
make: Entering directory '/home/walt/git/gravity/e/assets/robotest'
install={"installer_url":"/images/telekube-7.1.0-alpha.1.245.tar","nodes":3,"flavor":"three","role":"node","os":"ubuntu:18"}
install={"installer_url":"/images/opscenter-7.1.0-alpha.1.245.tar","nodes":1,"flavor":"standalone","role":"node","os":"ubuntu:18","ops_advertise_addr":"example.com:443"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.245.tar",""flavor":"three","nodes":3,"role":"node","os":"centos:7","storage_driver":"overlay2"}
resize={"installer_url":"/images/robotest-7.1.0-alpha.1.245.tar","nodes":1,"to":3,"flavor":"one","role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
shrink={"installer_url":"/images/robotest-7.1.0-alpha.1.245.tar","nodes":3,"flavor":"three","role":"node","os":"redhat:7"}
upgrade={"from":"/images/robotest-7.0.7.tar","installer_url":"/images/robotest-7.1.0-alpha.1.245.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.13.tar","installer_url":"/images/robotest-7.1.0-alpha.1.245.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.12.tar","installer_url":"/images/robotest-7.1.0-alpha.1.245.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.245.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
make: Leaving directory '/home/walt/git/gravity/e/assets/robotest'
```
</details>


<details><summary>nightly configs</summary>

```
walt@gcp:~/git/gravity$ make -C assets/robotest ROBOTEST_CONFIG=nightly get-config
make: Entering directory '/home/walt/git/gravity/assets/robotest'
install={"installer_url":"/images/telekube-7.1.0-alpha.1.255.tar","nodes":3,"flavor":"three","role":"node","os":"ubuntu:18"}
install={"installer_url":"/images/opscenter-7.1.0-alpha.1.255.tar","nodes":3,"flavor":"ha","role":"node","os":"ubuntu:18","ops_advertise_addr":"example.com:443"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"redhat:7","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"six","nodes":6,"role":"node","os":"redhat:7","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"six","nodes":6,"role":"node","os":"centos:7","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"debian:9","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"six","nodes":6,"role":"node","os":"debian:9","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"six","nodes":6,"role":"node","os":"ubuntu:16","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","storage_driver":"overlay2"}
install={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"six","nodes":6,"role":"node","os":"ubuntu:18","storage_driver":"overlay2"}
resize={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","to":3,"flavor":"one","nodes":1,"role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
resize={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","to":6,"flavor":"three","nodes":3,"role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
shrink={"installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","nodes":3,"flavor":"three","role":"node","os":"redhat:7"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"six","nodes":6,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"one","nodes":1,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.0.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-6.2.0.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-6.2.5.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-6.3.18.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.13.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.12.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"redhat:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"debian:9","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.16.tar","installer_url":"/images/robotest-7.1.0-alpha.1.255.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
make: Leaving directory '/home/walt/git/gravity/assets/robotest'
```
</details>

### Impact on the Build
This adds 20+GB of build artifacts to each PR build. Stats after a `clean` and `cache-clean` followed by `images`.

```
walt@gcp:~/git/gravity$ du -sh ~/.cache/robotest/ent e/build/current/robotest 
5.7G    /home/walt/.cache/robotest/ent
16G     e/build/current/robotest
```

Caches may eventually be shared between builds/branches, reducing that impact, but as of now they're kept in the build directory.

Robotest image builds adds ~7 minutes to master Gravity PR runtime.  With the new Jenkins pipeline, this can be easily seen as robotest image builds are split out into their own step.

Both of these numbers will adjust proportionally upwards for branches that have more upgrade base versions.  E.g. 5.5 has 10 upgrade base versions to master's 4.

### Jenkinsfile work

The refactored Jenkinsfile & Make can run both Nightly and PR configurations by tweaking the `ROBOTEST_CONFIG` variable.  After this PR merges, I will move nightlies to use the regular `Jenkinsfile`, and then retire `Jenkinsfile.nightly`.

To see what the new Jenkins pipleline looks like, see [the PR build for this PR](https://jenkins.gravitational.io/job/Gravity-PR-5.2/job/PR-2133/), or r[ecent try builds where I've been iterating on Jenkins changes](https://jenkins.gravitational.io/view/Robotest/job/gravity-try-build/).

r0mant committed:
Compare hostnames when searching for default hub credentials.

See https://github.com/gravitational/gravity/pull/2023.

(cherry picked from commit 63e0b219e8907d926e100c73d8e95b1a36069249)

wadells committed:
Revert "Revert "Remove `tele login` from automation. (#1890)""

This reverts commit 9981339b2d3b743388110096a8151d35e2614632.

wadells committed:
Refactor robotest to have its own Makefile.

With upgrade images (#1915) robotest is now performing relatively
complex build operations. We want to split this complexity from the
already complex gravity Makefiles.

wadells committed:
Collapse Robotest's run_suite.sh into build.assets/robotest/run.sh.

We used two different files to set environment variable defaults, check
files, and compose the correct docker command line.  These two are now
collapsed into one, for ease of editability, and deduplication of code.

wadells committed:
Remove -debug flag.

This is solely used for enabling the profiler, which shouldn't ever be
needed for CI and stable Gravity runs.

wadells committed:
Save robotest state with the current uid/gid.

Without this, state & logs are written as root:root, which means that
the build executor cannot delete them with its usual permissions.

Fixes #2103.

wadells committed:
Fix opscenter testing in new robotest Makefile.

opscenter.tar needs to end up in the /images directory within the docker
container -- this is where all images live now.

wadells committed:
Rename 'image' target to 'image-robotest'.

'image' vs 'images' was quite close, and thus confusing.

Also some cleanup of Makefile help text.

wadells committed:
WIP on image tarball naming.

wadells committed:
Use gravity version as robotest image version.

This avoids the following error during upgrade:

  [ERROR]: update version (0.0.0) must be greater than the currently installed version (0.0.0)

wadells committed:
Split current robotest application from upgrade base.

These need to be different to sufficiently exercise helm upgrades, as
well as other functionality where we want to migrate from an older code
path to a newer one.

wadells committed:
Remove extra leading space.

I'm not sure why these were here -- probably we could use some sort
of yaml linter in the repo.

wadells committed:
Use helm to upgrade from echoserver 1.0 to 1.10.

Name the helm release such that we can refer back to it from the
postupdate hook. Add a postupdate hook to run `helm upgrade`.

Decrease upgrade-base echosever image to 1.0, such that something
changes between the base and the current image.

Remove some hooks that won't be run on the upgrade-base image
(which is only installed and then upgraded away from -- no expands
or shrinks currently).

wadells committed:
Remove openebs from upgrade-base image.

6.2 (upgrade tested in nightlies) does not support this feature, and
fails with the following error:

  [ERROR]: could not parse the application manifest:
  I[#] S[#] doesn't validate with "schema.json#"
    I[#] S[#/$ref] doesn't validate with "#/definitions/Manifest"
      I[#] S[#/definitions/Manifest/additionalProperties]
  additionalProperties "storage" not allowed

wadells commented on assets/robotest/upgrade-base/charts/echoserver/Chart.yaml:
```diff
@@ -0,0 +1,4 @@
+name: echoserver
+version: 0.0.1
```
This should differ between `upgrade-base` and `current`.  Probably 1.0, 1.1.

wadells replied:
Changed in 3355ed7.

wadells commented on build.assets/robotest/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+ARG BASE
```
A comment & link to the relevant bug wouldn't go awry in this file.

wadells replied:
This entire file has been replaced by a simpler `--user=$(id -u):$(id -g)` in 
2e7bfe1

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+# With the advent of custom robotest images in upgrade testing (#1915)
+# constructing all the necessary cluster images became complex. As such
+# the logic has been quarantined in this file to reduce complexity in Gravity's
+# Makefile, e/Makefile, and build.assets/Makefile.
+#
+# Robotest image construction uses the following strategy:
+#
+#  - For the release under test, build using TELE_OUT and PACKAGES_DIR from
+#    the Gravity Makefiles.  This must occur after a successful `make telekube`
+#    (which populates PACKAGES_DIR with a base image). This Makefile does not
+#    know how to generate these and will fail if they're absent.
+#
+#  - For upgrade to/from versions:
+#
+#    1) Download & cache appropriate tele from our release infrastructure.
+#    2) Build & cache the robotest image using the current application manifest
+#       with the old tele. Build with a shared --state-dir caching released
+#       packages locally.
+#    3) Populate the current build directory from the image cache.
+#
+# Caching is critical because at the time of writing:
+#
+#  - Jenkins is pushing 20TB/mo of network data, primarily for robotest
+#    image construction & distribution.
+#  - Downloading one gravity base image is O(minutes). Multiply this by 3-10
+#    different versions used in upgrade testing and the build times become tedious.
+#
+# Misc additional concerns:
+#
+#  - All CACHE population must be atomic (file renames on the same filesystem). Otherwise
+#    a race between builds could result in a half copied artifact being used.
+#  - CACHES should not be assumed to be on the same filesystem as the build directories.
+#  - Enterprise tele and OSS tele cannot share caches, nor should their outputs
+#    be confused.
+#  - If any content in ROBOTEST_IMG_SRC changes, old images in ROBOTEST_UPGRADE_IMAGE_CACHE
+#    must be ruled out. These shouldn't necessarily be removed if the cache is
+#    shared between branches, as other branches may use the older images. In this
+#    scenario, ROBOTEST_TELE_CACHE and ROBOTEST_GRAVITY_PKG_CACHE remain valid.
+#  - ROBOTEST_IMAGEDIR can be heavy (up to 30GB for 7.0.x). This can
+#    aggravate CI workers that have many active builds or don't clean up.
```
Move this content to the README.

wadells replied:
Moved & integrated in f5a5aa3.

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+
+.PHONY: images
+images: ## Create all images necessary to run robotest CI tests.
+images: image-robotest images-upgrade image-opscenter
```
Consider doing this as a `$(MAKE) -j targets` to parallelize the main image and opscenter image work with the already parallel upgrade images -- alternatively, this may may the logs very difficult to follow.  If kept this way, leave a comment as to why.

wadells replied:
Parallelism won't work with a shared cache, as discussed in ecc6916.

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+get-upgrade-from-versions: ## Show versions used in upgrade from testing.
+get-upgrade-from-versions:
+	@echo $(ROBOTEST_UPGRADE_VERSIONS)
+export
```
export looks out of place

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+	$(MAKE) -j $(ROBOTEST_UPGRADE_CACHE_IMAGES)
+
+.PHONY: get-upgrade-from-versions
+get-upgrade-from-versions: ## Show versions used in upgrade from testing.
+get-upgrade-from-versions:
+	@echo $(ROBOTEST_UPGRADE_VERSIONS)
```
This should probably be `get config` and print out the rendered config.  I'm not sure there is much reason to care about the upgrade versions alone (aside from those hacking on this file).

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+
+.PHONY: run
+run: ## Run robotest.
+run: export GRAVITY_URL = $(ROBOTEST_GRAVITY_BINARY)
```
should probably be `?=` just for consistency.

wadells commented on build.assets/robotest/nightly_config.sh:
```diff
@@ -21,15 +21,21 @@ UPGRADE_MAP[6.2.0]="ubuntu:16"
 
 UPGRADE_VERSIONS=${!UPGRADE_MAP[@]}
 
+# The following breaks a dependency loop. We need upgrade versions to generate tarballs,
+# but we need tarball names to generate the full test config.
+if [[ ${1} == "upgradeexit" ]]  ; then
+    return
+fi
```
This still feels like an ugly hack.  How could the code be factored to handle this dependency better?

wadells commented on build.assets/robotest/pr_config.sh:
```diff
@@ -67,10 +72,11 @@ EOF
 }
 
-SUITE=$(build_install_suite)
+SUITE=""
+SUITE="$SUITE $(build_install_suite)"
```
This change should also occur in nightly.  (for easier commenting of these lines during development.

wadells commented on build.assets/robotest/pr_config.sh:
```diff
@@ -67,10 +72,11 @@ EOF
 echo "$SUITE" | tr ' ' '\n'
 
-export SUITE UPGRADE_VERSIONS
+export SUITE
```
Nightly needs this change too.

wadells replied:
Refactored nightly to include a newer version of this in 43c08e9.

wadells commented on build.assets/robotest/run.sh:
```diff
@@ -3,55 +3,114 @@ set -o errexit
+DESTROY_ON_SUCCESS=${DESTROY_ON_SUCCESS:-true}
+DESTROY_ON_FAILURE=${DESTROY_ON_FAILURE:-true}
+
+# set SUITE and ROBOTEST_IMAGE_DIR_MOUNTPOINT
```
`ROBOTEST_IMAGE_DIR_MONTPOINT` should come from this higher level instead of vice versa.

Also `IMAGES_DIR` instead of `IMAGE_DIR` for consistence with `ROBOTEST_IMAGES_DIR`

wadells replied:
Renamed in 40600dc49b149ad223bd7671aa3696da5a9abff4.

wadells commented on build.assets/robotest/run.sh:
```diff
@@ -3,55 +3,114 @@ set -o errexit
+EXTRA_VOLUME_MOUNTS="-v $ROBOTEST_IMAGES_DIR:$ROBOTEST_IMAGE_DIR_MOUNTPOINT"
+
+
+# GRAVTIY_FILE/GRAVITY_URL specify the location of the up-to-date gravity binary
+if [ -d $(dirname ${GRAVITY_URL}) ]; then
+  GRAVITY_FILE='/robotest-bin/'$(basename ${GRAVITY_URL})
+  EXTRA_VOLUME_MOUNTS=${EXTRA_VOLUME_MOUNTS:-}" -v "$(dirname ${GRAVITY_URL}):$(dirname ${GRAVITY_FILE})
+fi
```
File an enhancement bug to remove this dependency.  It should be quite unnecessary in this age.

wadells replied:
https://github.com/gravitational/robotest/issues/232

wadells commented on build.assets/robotest/run.sh:
```diff
@@ -3,55 +3,114 @@ set -o errexit
 }
 
-export EXTRA_VOLUME_MOUNTS=$(build_volume_mounts)
+if [ $DEPLOY_TO == "gce" ] ; then
```
DEPLOY_TO can't be anything else at this point. Consider refactoring everything related to this variable.

wadells commented on build.assets/robotest/run.sh:
```diff
@@ -3,55 +3,114 @@ set -o errexit
+installer_url: ${INSTALLER_FILE:-${INSTALLER_URL}}
+gravity_url: ${GRAVITY_FILE:-${GRAVITY_URL}}
+script_path: /robotest/terraform/${DEPLOY_TO}
+state_dir: /robotest/state
```
Turn the state_dir mount value into a variable.

wadells reviewed (commented):
First round of self review.

r0mant commented on assets/robotest/current/status.yaml:
```diff
@@ -0,0 +1,18 @@
+        - name: echoserver-status
+          image: quay.io/gravitational/debian-tall:buster
+          command:
+            - sh
```
Nit: I would use absolute path (`/bin/sh`) to avoid issues if the container doesn't have proper PATH set for some reason. Same for `wget`.

wadells replied:
Fixed in 2a82ca9.

r0mant commented on assets/robotest/current/status.yaml:
```diff
@@ -0,0 +1,18 @@
+          command:
+            - sh
+            - -c
+            - wget --quiet --output-document=- echoserver:$ECHOSERVER_SERVICE_PORT
```
Is this still going to print an error if the query fails even with the quiet flag set?

wadells replied:
Good point.  I chose quiet because of `--output-document=-`  Having wget's verbose (default) output interlaced was unsightly.  I changed this to `--no-verbose` in 2a82ca9.

r0mant commented on assets/robotest/upgrade-base/charts/echoserver/values.yaml:
```diff
@@ -0,0 +1,2 @@
+# echoserverRegistry is redefined to "leader.telekube.local:5000" in gravity hooks
```
`registry.local:5000`

wadells replied:
Changed in 900be40.

r0mant commented on assets/robotest/upgrade-base/app.yaml:
```diff
@@ -0,0 +1,158 @@
+apiVersion: cluster.gravitational.io/v2
+kind: Cluster
+metadata:
+  name: robotest
```
What do we need two robotest apps for, "upgrade-base" and "current"? To be able to test application-only upgrades when runtime doesn't change?

wadells replied:
We need two (or more) separate manifests for a couple reasons:

1) Old versions of gravity do not support all the manifest directives that current versions do.  For master (this PR) we test upgrades from:
```
6.2.0 6.2.5 6.3.18 7.0.0 7.0.12 7.0.13 7.0.16 (this last one floats)
```
In particular, 6.2.x doesn't support `storage/openebs`.  We can make the entirely reasonable argument that we don't need to test 6.2 for this branch.  However the mechanism is important because this code will be backported and we do need to test upgrades from 5.5.x in 7.0.x.  And there are manifest differences between 5.5..x and 7.1.x.  We need them to test contemporary features.  This is why I alluded to upgrade testing helping with backports yesterday.

2) In order to exercise the upgrade hook, we *want* the upgrade base application to be different than the current application. Ideally the hooks for the new version will exercise both gravity and application functionality that wasn't previously available. We're not there yet, but that is where I'm aiming.

To flip your question: Why don't we have more upgrade from versions?

Considering the list from earlier:
```
6.2.0 6.2.5 6.3.18 7.0.0 7.0.12 7.0.13 7.0.16 (this last one floats)
```

We could very well have one base image for 6.2.x and a different one for 6.3 and 7.0.x. Or a different version of the application for each upgrade from version, supporting exactly what that gravity version supports.

There are a couple difficulties here:

1) Maintaining a separate application for each upgrade base version is tedious. It means each manifest change is multiplied by 5-10 edits.  There is potential to put together some code to only abstract & maintain the differences.  But this seemed complex for a v1.  I'm not looking to build out yaml/manifest templating.

2) There is no good way to keep a manifest current for the floating version. It'd require that each time you cut a release, you (or some automation) cut a PR to each branch that could be upgraded to/from.  This is not worth it at this point.

2) Makefiles are insufficient to elegantly handle this.  The pattern rules I've written are already byzantine and difficult to understand, without having many-to-one or one-to-one manifest-to-upgrade-images.  This would be considerably easier using go & @knisbet's work in #1917 but I can't afford to tie my work to his -- it'd  delay availability by weeks-to-months and make ports... challenging.

a-palchikov replied:
Just to chime in on the topic of yaml templating - if it's ever a thing, I think [cue](github.com/cuelang/cue/) would be perfect for the job.

r0mant commented on build.assets/robotest/nightly_config.sh:
```diff
@@ -21,15 +21,21 @@ UPGRADE_MAP[6.2.0]="ubuntu:16"
 
+# The following breaks a dependency loop. We need upgrade versions to generate tarballs,
+# but we need tarball names to generate the full test config.
+if [[ ${1} == "upgradeexit" ]]  ; then
```
I'm not sure what this is for. Could you explain in more detail?

wadells replied:
I added extra documentation on these config files & how they're used in ccff414d9431d0e51400df43114baa851c31d8ee.  Check out `assets/robotest/config/README.md` and see if it sufficiently clarifies the implementation.

r0mant commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+$(ROBOTEST_TELE_CACHE)/tele-%: | $(ROBOTEST_CACHE_TMP) $(ROBOTEST_TELE_CACHE)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+ifdef ENTERPRISE
+	wget --quiet https://get.gravitational.io/telekube/bin/$*/linux/x86_64/tele --output-document $(TMP)/tele
```
Can we find a way to avoid placing enterprise-specific URLs in the OSS code? For example, instead of setting `ENTERPRISE` variable in the enterprise code, set `TELE_URL` or something like this and use it here.

wadells replied:
The issue here is that OSS gravity doesn't provide a direct link to download tele that I've found - we need to download it as part of a tarball, unpack and then copy just the `tele` binary.  Whereas for enterprise we need to download just the binary, or authenticate to get.gravitational.io to get a release tarball.  I haven't been able to find a common download path.

If we wanted to properly abstract this difference in behavior, we'd need to either:
* Find a common download path.  Two unauthenticated release tarballs or two unauthenticated raw binaries.
* Declare & call a function pointer, if such a thing is even possible in a Makefile. This seems like an poor choice from a maintenance perceptive.
* Do some sort of `sed`-esque replacement of a `TELE_URL` variable to plug in the variety of upgrade-from versions.

Let me know what you think.  The value of keeping enterprise stuff out of OSS stuff is clear to me, but the path to doing so is not.

a-palchikov replied:
I can imagine having some e-specific make functions stored in e/ which could be conditionally included with:
```
-include e/makefile.conf
```
Then we can factor the tele pull process into a function that accepts the version and output path:
```
define pull_tele
	wget --quiet https/path/to/telekube/bin/$1/linux/x86_64/tele --output-document $2
	chmod u+x $2
endef
```

wadells replied:
Thanks @a-palchikov.  I'll run with your suggestion.

wadells replied:
I've got a working version of this change in place as of c91653e.

I use a shell script instead of a included Makefile to accommodate tmpdir cleanup per https://github.com/gravitational/gravity/pull/2133#discussion_r489093866.

I want to make one more structural change in this area before calling it done.  That is:

Migrate all content from `build.assets/robotest` to `assets/robotest`.  This will give a nice parity between the enterprise `assets/robotest` and the oss `assets/robotest`.

wadells replied:
Take a look at `assets/robotest/download_tele.sh` in the two repos and see if the new implementation suits your fancy.

r0mant commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+# .PRECIOUS because it is a cache shared across versions.
+.PRECIOUS: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar
+$(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar: $(ROBOTEST_TELE_CACHE)/tele-% $(ROBOTEST_UPGRADE_BASE_IMAGE_SRC) | $(ROBOTEST_UPGRADE_IMAGE_CACHE) $(ROBOTEST_CACHE_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
```
I wonder if we should consider trapping `rm -rf $(TMP)` to make sure that the directory is removed even if, say, tele build below fails. Same for other targets that create temp. dirs.

wadells replied:
Makefiles have awful signal handling.  We'd need to shell out to bash.  This is possible, but it is adding yet another layer of complexity.  I'd considered this, but was hoping that we could wait until mage/go builds, where signal handling and defer/recover is first class.  Especially with the newly enlarged jenkins workspace giving us buffer for heavy PRs

As of this PR both `$(ROBOTEST_CACHE_TMP)` and `$(ROBOTEST_BUILD_TMP)` are underneath the build root.  The caches, along with any straggling temp files will be nuked by a `make clean`. If the user puts `$(ROBOTEST_CACHE_ROOT)` outside the build dir, leaking tmp directories is a concern -- but that isn't the case for jenkins (yet).  It is more of a concern for local development.  I'd like to defer this until either:

* We move `$(ROBOTEST_CACHE_TMP)` outside `$(BUILDDIR)` or
* We migrate to mage/go for our build files

wadells replied:
I've refactored out all the tmp stuff into shell scripts with traps at of c91653e. Although I think ithe code is a bit more difficult to undestand, it aligned nicely with containerizing the `tele build` which was needed to handle `~/.tsh` pollution across a large range of `tele`s.

wadells commented on assets/robotest/current/install.yaml:
```diff
@@ -0,0 +1,24 @@
+    metadata:
+      name: echoserver-install
+    spec:
+      restartPolicy: Never
```
If the helm install fails, we have no reason to believe it would succeed if run again.

wadells replied:
While I was developing this, when `helm install` failed I found no reason to believe it would succeed when run a second, third, or fourth time.

`OnFailure` resulted in failures on an exponential backoff basis, resulting in longer timeouts before failure and a slower development cycle.  Perhaps if I knew more about the failure modes here I could make a more educated choice about whether to abort or retry.

r0mant replied:
> when helm install failed I found no reason to believe it would succeed when run a second, third, or fourth time

Technically that's not entirely true, helm install can fail for example due to temporary etcd issues and succeed on the retry.

wadells replied:
Changed to `OnFailure` in 43c65ee.

wadells commented on Makefile:
```diff
@@ -530,28 +522,13 @@ $(TELEKUBE_OUT): packages
 		--skip-version-check \
 		-o $(TELEKUBE_OUT)
 
-$(GRAVITY_OUT):
```
I removed the code I added in my previous custom robotest image PR:

https://github.com/gravitational/gravity/pull/1967/files#diff-b67911656ef5d18c4ae36cb6741b7965R533-R537

Is your use of `make gravity tele` more recent than that?  I can definitely add these back in.

I added this code because robotest targets depend on $(GRAVITY_OUT) and $(TELE_OUT).  They used to explicitly do so, but this was unpleasant because these tele and gravity depend on phony targets... meaning they constantly trigger a full rebuild.  For the new `robotest-run-suite` and `robotest-run-nightly`, in this PR I implicitly declare the dependencies in a comment to avoid this.  Maybe these should be an "ordering dependency"  Its not great... but I also don't want to fix our entire dependency tree in this PR -- doing so would be a lot of work and orthogonal to the changes here.

wadells replied:
This shouldn't affect `make gravity tele`, that uses the `$(BINARIES)` target.  `$(GRAVITY_OUT)` and `$(TELE_OUT)` are the full paths to the binaries, not the .PHONY targets you're calling.

wadells committed:
Fix cache directory when nested under $(BUILDDIR).

No need to include oss/ent when that is already implicit in the
builddir.

wadells committed:
Separate tele build logic into a shell wrapper.

This makes it more difficult to leak partially constructed images in
temp directories is a build is called off mid run (e.g. with ctrl-C).

wadells committed:
Isolate robotest tele build from ~/.tsh.

Shared ~jenkins state strikes again. We need to isolate enterprise
robotest image builds from ci-ops login credentials stored in ~/.tsh.

Without this, enterprise builds use the most recent opscenter that
was `tele login`'d, resulting in unofficial and missing artifacts
when ci-ops is chosen, as described in:

  https://github.com/gravitational/gravity/pull/2133#discussion_r489075451

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+$(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar: $(ROBOTEST_TELE_CACHE)/tele-% $(ROBOTEST_UPGRADE_BASE_IMAGE_SRC) | $(ROBOTEST_UPGRADE_IMAGE_CACHE) $(ROBOTEST_CACHE_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+	$(eval TGT := $(TMP)/robotest-$*.tar)
+	$< build \
```
Should be fixed by 5c47ccf. :crossed_fingers:

wadells replied:
Unfortunately 5c47ccf failed because not all upgrade from `tele` binaries have the fix from #2023, leading to `[ERROR]: no credentials for https://get.gravitational.io:443` if `--hub` is blanket applied to old tele builds.

I'd rather not branch the upgrade base tele build logic at this point, so I've opted to containerize all tele builds to isolate them from ~jenkins state.  We'll see how this goes shortly.

wadells replied:
Refactored to sandbox tele builds within a docker container in c91653e.  This prevents any possible contamination from ~jenkins.  It took a couple more patches to get the new dockerized builds working, but now that they are (f86dc73) we don't see this issue any more.

wadells commented on build.assets/robotest/Makefile:
```diff
@@ -0,0 +1,326 @@
+$(ROBOTEST_IMAGE): $(TELE_OUT) $(ROBOTEST_IMAGE_SRC) | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval TGT := $(TMP)/robotest-$(GRAVITY_VERSION).tar)
+	$(TELE_OUT) build \
```
Should be fixed by 5c47ccf. crossed_fingers

a-palchikov commented on build.assets/robotest/tele_build.sh:
```diff
@@ -0,0 +1,31 @@
+#
+# This file provides the following features to the tele build process:
+#
+#  1) Atomicity.  The TARGET is either complete or absent. (So long as long as
```
Either `So long` or `as long as`

wadells replied:
Fixed in cbb8da7.

a-palchikov commented on build.assets/robotest/tele_build.sh:
```diff
@@ -0,0 +1,31 @@
+# TELE  - The tele to build the image, should typically be equal to VERSION
+# VERSION - The application version to use in the cluster image
+# APP_YAML - The image manifest to be built.
+# STATE_DIR - The gravity state dir where packages are cached/drawn from. May be a shared with parallel invocations.
```
~a~ shared?

wadells replied:
Fixed in cbb8da7.

a-palchikov commented on build.assets/robotest/tele_build.sh:
```diff
@@ -0,0 +1,31 @@
+
+TMP=$(mktemp -d -p $BUILD_TMP)
+trap "rm -rf $TMP" exit
+TGT=$TMP/$(basename $TARGET)
```
can this have whitespace?

wadells replied:
I added a bunch of quoting in 6deb147.  I think I caught all the possible filepaths.

a-palchikov commented on build.assets/robotest/tele_build.sh:
```diff
@@ -0,0 +1,31 @@
+trap "rm -rf $TMP" exit
+TGT=$TMP/$(basename $TARGET)
+
+$TELE build \
```
I'd also take into account that paths can contain whitespace.

wadells replied:
I added a bunch of quoting in 6deb147.  I think I caught all the possible filepaths.

wadells committed:
Quote filepaths that may have spaces in them.

wadells committed:
Remove upgrade image cache.

Sharing upgrade images between parallel invocations is complex,
different branches will have different manifests and we need a way to
ensure that tarballs are not incorrectly recycled. Best to postpone this
until further design is complete.

wadells committed:
Refactor robotest image builds for enterprise.

Most noteably:

 - Sandbox tele build in a container to keep it from reading ~jenkins/.tsh
 - Provide a new mechanism for downloading tele binaries.

wadells committed:
Create gravity state dir before mounting to docker.

Without this, docker will create the directory with root permissions,
which will subsequently prevent tele build from using it.

wadells committed:
Update tele mtime before moving into place.

When untaring, the mtime is preserved from when the file was saved into
the tar.  This means it will usually predate the dependencies, and cause
the target to be continually rebuilt.  This is fixed by updating the
mtime before moving the file into place.

wadells committed:
Serialize upgrade base image builds.

Having these parallelized was resulting in package sync errors on
jenkins. Noteably I was seeing a 100% repro rate for 7.0.16 builds
hitting an error like the following:

  [ERROR]: failed to sync packages for runtime version 7.0.16
      package(gravitational.io/storage-app:0.0.3) already exists

The exact package varied.

wadells committed:
Replace telekube.leader.local with registry.local.

This is the preferred nomeclature, per:

  https://github.com/gravitational/gravity/pull/2133#discussion_r489080546

wadells committed:
Use full path to binaries in containers.

Just in case we get a container with a broken PATH.

Also, swapped --quiet for --no-verbose, to give a bit of useful output
upon failure.

wadells committed:
Use touch instead of --iidfile.

--iidfile is not available on our build infrastructre currently

See https://github.com/gravitational/ops/issues/141.

wadells committed:
Split robotest Jenkinsfile configs.

Now that robotest has its own substantive build steps, lets clean up the
pipeline a bit by factoing out robotest related code.

wadells committed:
Use the tele/gravity version as the upgrade base version.

Using $(GRAVITY_VERSION) caused all images to have the same version,
preventing upgrades.

wadells commented on assets/robotest/current/postupdate.yaml:
```diff
@@ -0,0 +1,22 @@
+            - /usr/local/bin/helm
+            - upgrade
+            - --set
+            - echoserverRegistry=leader.telekube.local:5000
```
Changed in 900be40.

wadells commented on assets/robotest/current/install.yaml:
```diff
@@ -0,0 +1,24 @@
+            - --name
+            - echoserver
+            - --set
+            - echoserverRegistry=leader.telekube.local:5000
```
Changed in 900be40.

wadells committed:
Fix typos.

Per:
  https://github.com/gravitational/gravity/pull/2133#discussion_r489624075
  https://github.com/gravitational/gravity/pull/2133#discussion_r489624876

wadells committed:
Simplify setting robotest file ownership via docker.

After working with containerizing tele build and inspecting our other
make targets, the `--user` flag is a much easier way to set log
ownership than baking a whole new image with a custom user inside.

wadells committed:
Ensure all paths are fully qualified.

This greatly simplifies enterprise integration by allowing the oss
makefile to 'include'd directly, instead of reimplementing targets.

Culled some unused variables and clarified the use of others.

wadells committed:
Add a no-op robotest config for makefile testing.

This quick to build & quick to run configuration allows verification of
the build infrastructure without building a bunch of images or lengthy
cloud provisioning & gravity install.

Also, several phony targts have been updated to correctly do the correct
thing when they have zero dependencies -- which can happen if we don't
need any upgrade images for the configuration.

wadells committed:
Move config scripts into a dedicated directory.

This simplifies config management, and gives a convienient place for a
README or other documentation on the config scripts.

wadells committed:
Update paths to accommodate config directory.

Refactor how config scripts pass configurations to the Makefile.

Add a README describing use of config scripts.

wadells committed:
Merge build.assets/robotest into assets/robotest directory.

No need to have both now.  build.assets/robotest was a legacy of a
different era of robotest build & execution. This also gives a nice
symmetry to the enterprise directory layout.

wadells committed:
Adjust paths to reflect a Makefile inside the directory.

wadells committed:
Move robotest credentials into the robotest directory.

This consolidates robotest files in one location.

wadells committed:
Consolidate robotest documentation in README.md.

To help folks get started developing robotest images and configurations
and reduce the amount of docs maintenance, all docs have been
consolidated in the README.md files.

wadells commented on .gitignore:
```diff
@@ -18,3 +18,5 @@ gravity.log
 # Ansible retry files
 *.retry
+# Robotest credentials
+robotest-gcp-credentials.json
```
Moved into the robotest directory in dfe9106.

wadells committed:
Differentiate upgrade-base and current robotest images further.

In particular, we want to have different helm chart versions to make
sure that that the helm upgrade actually moves from an older version of
the release to a newer version.

wadells committed:
Update e - fix enterprise robotest variables expansion.

Several variables were being lazily expanded after TOP/ROOT had been
changed to their OSS values.

wadells commented on Jenkinsfile:
```diff
@@ -33,69 +33,85 @@ def setRobotestParameters() {
+    stage('run robotest') {
+      if (runRobotest) {
+        withCredentials([
+					file(credentialsId:'ROBOTEST_LOG_GOOGLE_APPLICATION_CREDENTIALS', variable: 'GOOGLE_APPLICATION_CREDENTIALS'),
```
Looks like use of spaces vs tabs is inconsistent in this file. Probably need to install a Groovy indent plugin.

wadells replied:
Fixed in f3a02f8.

wadells commented on assets/robotest/README.md:
````diff
@@ -0,0 +1,165 @@
+export ROBOTEST_GOOGLE_APPLICATION_CREDENTIALS=~/.config/gcloud/legacy_credentials/robotest.json
+```
+
+Once these are set, robotest with:
````
*run* robotest with

wadells replied:
Fixed in 593ede2.

wadells committed:
Fix docs errors.

wadells committed:
Re-indent Jenkinsfile

wadells committed:
Add a link to `tele build` parallelism bug #2060.

wadells committed:
Fix a typo in a variable name.

Unfortunately, this typo caused all upgrades to be skipped.

wadells committed:
Add telekube testing back into pr and nightly configs.

With the advent of robobotest specific images, all testing was migrated
to the robotest images.  This meant we no longer tested the telekube.tar
produced by the build. This PR reintroduces telekube.tar testing because
this is an artifact we distribute to customers/users.

wadells committed:
Do not recursively expand any $(shell ...) calls.

Doing so results in needless re-execution of the same shell command,
resulting in noticeable initialization lag.

wadells committed:
Clarify stage name.

unittest is a python thing.  Unit test is the generally accepted way to
write the term.

r0mant commented on assets/robotest/config/pr.sh:
```diff
@@ -59,18 +64,25 @@ function build_install_suite {
-echo "$SUITE" | tr ' ' '\n'
-
-export SUITE UPGRADE_VERSIONS
+if [[ ${1} == "upgradeversions" ]] ; then
```
Maybe not in this PR but I would seriously consider moving all of this logic for generating test suite configurations into a Go program so the only shell/makefile code we have is a simple harness for building tarballs and launching robotest. Just don't want to end up in a situation with a mountain of non-trivial shell scripts to maintain.

How hard do you think it would be to do?

wadells replied:
I couldn't agree more.  These old shell scripts (that I've evolved considerably) are difficult to follow & debug. I don't think it would be hard.  I was actually tempted to rewrite in go it for this PR, but it felt like unnecessary scope creep.

wadells replied:
I've filed https://github.com/gravitational/gravity/issues/2147 tracking this work item and will triage it as I asses next workitems for https://github.com/gravitational/gravity/projects/5

wadells committed:
Change install hook restartPolicy to OnFailure.

Helm install may fail for reasons that self recover (e.g. etcd
troubles). As such, we prefer the faster timeout within this job to the
slower timout of robotest reprovision & retry. As discussed at:

  https://github.com/gravitational/gravity/pull/2133#discussion_r489080401

wadells committed:
Fix a typo.

wadells commented on assets/robotest/tele_build.sh:
```diff
@@ -0,0 +1,53 @@
+#
+# The following environment variables must be specified by the caller:
+#
+# IMAGE - A docker cotainer that the build will run in
```
Fixed in 32dd236.

a-palchikov commented on assets/robotest/Makefile:
```diff
@@ -0,0 +1,306 @@
+# ROBOTEST_UPGRADE_BASE_IMAGE_SRCDIR is the directory containing all sources to build the
+# robotest upgrade base application. It is mounted into a build container.
+ROBOTEST_UPGRADE_BASE_IMAGE_SRCDIR = $(TOP)/upgrade-base
+# ROBOTEST_UPGRADE_BASE_IMAGE_SRC is the gravity application manifest to build into robotest images.
```
from the implementation it looks as a list of files in the `$ROBOTEST_UPGRADE_BASE_IMAGE_SRCDIR` directory - copy/paste comment?

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/Makefile:
```diff
@@ -0,0 +1,306 @@
+# build all robotest upgrade images.
+ROBOTEST_TELE_BINARIES := $(addprefix $(ROBOTEST_TELE_CACHE)/tele-, $(ROBOTEST_UPGRADE_VERSIONS))
+# ROBOTEST_UPGRADE_IMAGES is the list of build specific upgrade images, suitable
+# for mounting into the robotest docker container. These are not shared stated.
```
stated -> state?

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/Makefile:
```diff
@@ -0,0 +1,306 @@
+# Because we must support old tele binaries to build old image versions, these races seem
+# unavoidable.
+#
+# Interestingly, I've never seen sync races happen with oss tele and the s3Syncer
```
This is b/c oss syncer implementation always downloads into an abritrary temporary directory.

a-palchikov commented on assets/robotest/README.md:
```diff
@@ -0,0 +1,165 @@
+
+### Running Robotest
+
+Before runnning robotest, set the following environment variables needed to
```
runn~~n~~ing

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/README.md:
```diff
@@ -0,0 +1,165 @@
+For upgrade to/from versions:
+
+ 1. Download & cache appropriate tele from the gravity release infrastructure.
+ 2. Build the the robotest image using the `upgrade-base` manifest
```
Build the ~~the~~

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/README.md:
```diff
@@ -0,0 +1,165 @@
+### Using Alternative Configurations
+
+By default, all targets use the `pr` configuration found in the `config/`
+directory. A `noop` configuration this does not rely on any upgrade images is
```
this -> that

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/README.md:
```diff
@@ -0,0 +1,165 @@
+## File Structure
+
+ * `Makefile` is the entrypoint. Run `make help` to see targets.
+ * `config/` contain branch specific suite configurations. See the README.md therein.
```
contain**s**

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/config/README.md:
```diff
@@ -0,0 +1,41 @@
+ * INSTALLER_URL
+
+`./script.sh configuration` must return a robotest suite configuration as described in
+https://github.com/gravitational/robotest/blob/master/suite/README.md.  For instance:
```
nit: I would use a specific tag/commit to refer to the suite README.md if possible. I keep running in situations with other repositories where links go stale pretty fast.

wadells replied:
Pinned in dcc3d40.

a-palchikov commented on assets/robotest/download_tele.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+#
+# This file is used to download tele binaries.  It provies the following benefits:
```
provides

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/download_tele.sh:
```diff
@@ -0,0 +1,28 @@
+#
+# This file is used to download tele binaries.  It provies the following benefits:
+#
+#  1) Atomicity.  The TARGET is either complete or absent. (So long as long as
```
`So long` or `as long as`

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/download_tele.sh:
```diff
@@ -0,0 +1,28 @@
+#  1) Atomicity.  The TARGET is either complete or absent. (So long as long as
+#     BUILD_TMP is on the same filesystem as TARGET and that fs supports atomic
+#     renames). This means no half finished files for Make to trip on.
+#  2) The file can be overridden/replaced in the make file to allow enterprise
```
nit: make file -> Makefile

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/download_tele.sh:
```diff
@@ -0,0 +1,28 @@
+wget --no-verbose "https://get.gravitational.com/gravity-$VERSION-linux-x86_64-bin.tar.gz" --directory-prefix "$TMP"
+tar -xzf "$TMP/gravity-$VERSION-linux-x86_64-bin.tar.gz" -C "$TMP"
+# touch because unpacking from gnu tar keeps the original mtime
+touch "$TMP/tele"
```
I'm pretty sure tar supports `--touch`

wadells replied:
Changed in c2a13c4.

a-palchikov commented on assets/robotest/run.sh:
```diff
@@ -0,0 +1,118 @@
+readonly TARGET=${1:?Usage: [path to config]}
+
+GRAVITY_URL=${GRAVITY_URL:?Set GRAVITY_URL to the current gravity binary}
+IMAGEDIR_MOUNTPOINT=${IMAGEDIR_MOUNTPOINT:? Set IMAGEDIR_MOUNTPOINT to the mountpoint where imags will be availabe inside the container}
```
imag**e**s

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/run.sh:
```diff
@@ -0,0 +1,118 @@
+IMAGEDIR_MOUNTPOINT=${IMAGEDIR_MOUNTPOINT:? Set IMAGEDIR_MOUNTPOINT to the mountpoint where imags will be availabe inside the container}
+INSTALLER_URL=${INSTALLER_URL:?Set INSTALLER_URL to the container local default robotest image}
+OPSCENTER_URL=${OPSCENTER_URL:?Set OPSCENTER_URL to the container local opscenter image}
+TELEKUBE_URL=${TELEKUBE_URL:?Set TELEKUBE_URL to the container local opscenter image}
```
opscenter -> telekube

wadells replied:
Fixed in d2cfbf4.

a-palchikov commented on assets/robotest/run.sh:
```diff
@@ -0,0 +1,118 @@
+export INSTALLER_URL TELEKUBE_URL OPSCENTER_URL IMAGEDIR_MOUNTPOINT
+SUITE=$($TARGET configuration)
+
+# GRAVTIY_FILE/GRAVITY_URL specify the location of the up-to-date gravity binary
```
GRAV**IT**Y_FILE

wadells replied:
Fixed in d2cfbf4.

a-palchikov reviewed (approved):
This is great work, Walt.
lgtm with just a couple of typos.

wadells commented on Jenkinsfile:
```diff
@@ -7,95 +7,111 @@ def propagateParamsToEnv() {
+    }
+  }
+  runRobotest = (env.RUN_ROBOTEST == 'run')
+    stage('build robotest images') {
```
I'm not fully sure I understand what you're asking.

I put the build in a separate step to help tracking timing & visualization.  The `run` target will build all images necessary if they're not already complete -- and we could call it directly without this pre-bui.d But I found it useful see whether building images or running robotest was the problem.

a-palchikov replied:
No, it just looked as if `runRobotest = ...` was starting a new scope where the following `stage` was nested. I just meant it could be on the same indentation level:
```
runRobotest = (env.RUN_ROBOTEST == 'run')
stage('build robotest images') {
...
```

wadells replied:
Thanks! Fixed in 4720bf4.

wadells commented on Makefile:
```diff
@@ -530,28 +522,13 @@ $(TELEKUBE_OUT): packages
-
-ROBOTEST_TAR_SRC=$(shell find $(ASSETSDIR)/robotest/ -type f)
+.PHONY: robotest-run-suite
+robotest-run-suite: # depends on: telekube opscenter $(TELE_OUT) $(GRAVITY_OUT)
```
Yes.

These .`.PHONY` upstream targets don't track dependency relations between sources and artifacts, meaning if `robotest-run-suite` depends on them directly it will always rebuild `production telekube opscenter`. This is tedious for anything but an automated builder that does it from scratch each time.  `robotest-run-suite` used to not declare any dependencies (which is a lie) -- and these commented out dependencies are my compromise.

We might similarly ask why all the `packages` targets don't depend on `$(GRAVITY_OUT)`? Or why doesn't  `$(TELE_OUT)` doesn't depend on `$(TELE_OUT)`?  They should, but to do that we need proper dependency relations between `$(GRAVITY_OUT)` and `$(TELE_OUT)` and their upstreams.  This quickly balloons into fixing the entire build.

I felt fixing latent broken dependency relationships  was out of scope. Especially as make's timestamp based tracking leaves a lot to be desired.  If we're going to invest this sort of energy into the build, it should be coupled with a modern build system IMO.

a-palchikov replied:
Agree - we should better focus on mage progress instead of fixing makefiles.

wadells commented on assets/robotest/Makefile:
```diff
@@ -0,0 +1,306 @@
+PACKAGES_DIR ?= $(GRAVITY_BUILDDIR)/packages
+# ROBOTEST_DOWNLOAD_TELE_SCRIPT allows the e Makefile to override how tele binaries are sourced.
+ROBOTEST_DOWNLOAD_TELE_SCRIPT ?= $(TOP)/download_tele.sh
+# ROBOTEST_CACHE_FLAVOR may be overriden to specify a non-compatible cache.
```
It is just the name of the cache sub directory.  I was having issues with a long-lived cache that lived outside `$(GRAVITY_BUILDDIR)` without this -- oss teles & enterprise teles were getting confused.

wadells replied:
I updated the comment to hopefully clarify this usage in d2cfbf4.

wadells committed:
Fix indentation.

wadells committed:
Use `tar --touch` instead of a `touch` later.

Per the recommendation here:
  https://github.com/gravitational/gravity/pull/2133#discussion_r493807326

wadells committed:
Fix a plethora of typos.

wadells committed:
Pin floating master ref at v2.1.0.

This prevents the link from breaking if this content is moved around.

wadells commented on assets/robotest/Makefile:
```diff
@@ -0,0 +1,306 @@
+#   $(HOME)/.gravity
+ROBOTEST_GRAVITY_PKG_CACHE ?=  $(ROBOTEST_CACHEDIR)/gravity-state
+# ROBOTEST_CACHE_TMP is a staging directory for building/copying/unpacking content
+# before an atomic rename into the. Must be on the same filesystem as ROBOTEST_TELE_CACHE
```
Fixed in d2cfbf4.

wadells commented on assets/robotest/Makefile:
```diff
@@ -0,0 +1,306 @@
+# parallel robotest builds in CI infra.
+ROBOTEST_TELE_CACHE ?= $(ROBOTEST_CACHEDIR)/tele
+# ROBOTEST_CACHES are all the directories on this machine used to speed up
+# upgrade artifact contruction by avoiding redownloading/rebuilding released
```
Fixed in d2cfbf4.

wadells committed:
Fix a call to an old function.

Ops is now handled as a top level call instead of part of the install
suite.

wadells committed:
Update e ref.

This picks up a couple late breaking fixes in the enterprise build
logic.
